contracts/src/challengeV2/libraries/EdgeChallengeManagerLib.sol

```diff
@@ -469,6 +469,55 @@ library EdgeChallengeManagerLib {
         return (hasRival(store, edgeId) && store.edges[edgeId].length() == 1);
     }
 
+    function updateAccuTimerCache(EdgeStore storage store, bytes32 edgeId, bytes32 claimingEdgeId, uint8 numBigStepLevel) internal returns (uint64) {
+        // calculate the time unrivaled without inheritance
+        // we don't use timeAccumulated here, since it would read the cache and double-count
+        uint64 accuTimer = timeUnrivaled(store, edgeId);
+        if (store.edges[edgeId].lowerChildId != bytes32(0)) {
+            uint64 lowerTimer = timeAccumulated(store, store.edges[edgeId].lowerChildId, numBigStepLevel);
+            if (lowerTimer == type(uint64).max) {
+                return type(uint64).max;
+            }
+            uint64 upperTimer = timeAccumulated(store, store.edges[edgeId].upperChildId, numBigStepLevel);
+            if (upperTimer == type(uint64).max) {
+                return type(uint64).max;
+            }
+            accuTimer += lowerTimer < upperTimer ? lowerTimer : upperTimer;
+        } else {
+            checkClaimIdLink(store, edgeId, claimingEdgeId, numBigStepLevel);
+            uint64 claimTimer = timeAccumulated(store, claimingEdgeId, numBigStepLevel);
+            if (claimTimer == type(uint64).max) {
+                return type(uint64).max;
+            }
+            accuTimer += claimTimer;
+        }
+        uint64 currentAccuTimer = store.edges[edgeId].accuTimerCache;
+        if (accuTimer > currentAccuTimer) { // only update when increased
+            store.edges[edgeId].accuTimerCache = accuTimer;
+            return accuTimer;
+        }
+        return currentAccuTimer;
+    }
+
+    function timeAccumulated(EdgeStore storage store, bytes32 edgeId, uint8 numBigStepLevel) internal view returns (uint64) {
+        uint64 accuTimerCache = store.edges[edgeId].accuTimerCache;
+        if (accuTimerCache > 0){
+            return accuTimerCache;
+        }
+        if (store.edges[edgeId].status == EdgeStatus.Confirmed 
+                && store.edges[edgeId].length() == 1
+                && ChallengeEdgeLib.levelToType(store.edges[edgeId].level, numBigStepLevel) == EdgeType.SmallStep) {
+            return type(uint64).max;
+        }
+        uint64 accuTimer = timeUnrivaled(store, edgeId);
+        if (store.edges[edgeId].lowerChildId != bytes32(0)) {
```

Comment:
We shouldn't need to do this recursively if each royal edge just calls `updateAccuTimerCache` right? If done from the very bottom of an OSP edge to the top, it the cache should be set for each edge

Reply:
yes, it will just break if cache exists; you might not want to cache everything because it is cheaper not to (SSTORE is expensive)